packages/core/src/types.ts

```diff
@@ -1221,25 +1220,39 @@ export namespace SCXML {
 }
 
 // Taken from RxJS
-export interface Unsubscribable {
+export interface Observer<T> {
+  next: (value: T) => void;
+  error: (err: any) => void;
+  complete: () => void;
+}
+
+export interface Subscription {
   unsubscribe(): void;
 }
+
 export interface Subscribable<T> {
+  subscribe(observer: Observer<T>): Subscription;
   subscribe(
-    next?: (value: T) => void,
+    next: (value: T) => void,
     error?: (error: any) => void,
     complete?: () => void
-  ): Unsubscribable;
-}
-
-export interface Observer<T> {
-  next: (value: T) => void;
-  error: (err: any) => void;
-  complete: () => void;
+  ): Subscription;
 }
 
 export type Spawnable =
   | StateMachine<any, any, any>
   | Promise<any>
   | InvokeCallback
   | Subscribable<any>;
+
+export interface ActorRef<TEvent extends EventObject, TEmitted = any>
+  extends Subscribable<TEmitted> {
+  send: Sender<TEvent>;
+  [key: string]: any;
```

Comment:
In general, it would be of great help to me (and probably to the users as well) if we had some kind of a page documenting our approach towards TS types. For instance, I try to stay away from `any` (as defaults for generics, object values, indexer values, etc) as much as I can. IMHO it's better to be safe by default and cast things where needed if necessary as it at least "documents" where "unsafe boundaries" lie in the code. The codebase is often not written in such a way so I naturally (to me) comment on each new call introducing such things - but if more explicit design goals would be written down then I could maybe just think more in their spirit.

Reply:
That's true, I'll omit `[key: string]` and make it more explicit.

> but if more explicit design goals would be written down then I could maybe just think more in their spirit.

Yeah, we should do this. V5 should definitely have some design goals re: TypeScript, and I'm sure @mattpocock can help with this.